ocaml/typing/typecore.ml

```diff
@@ -3348,7 +3367,10 @@ let rec is_nonexpansive exp =
       is_nonexpansive e &&
       List.for_all
         (fun {c_lhs; c_guard; c_rhs} ->
-           is_nonexpansive_opt c_guard && is_nonexpansive c_rhs
+           let guard_exp = Option.map (function
+             | Typedtree.Predicate p -> p
+             | Typedtree.Pattern (e, _, _) -> e) c_guard in
+           is_nonexpansive_opt guard_exp && is_nonexpansive c_rhs
```

Comment:
I wonder if we also need to check whether the pattern in the pattern guard contains an exception pattern.

Reply:
makes sense, I added this check

Reply:
This looks right, thanks. I would vaguely prefer to extract out both `e` and `pat` from the same pattern match, so it's more visually clear that we're considering all bits of information we ought to. That might look like:

```ocaml
let is_guard_nonexpansive =
  match c_guard with
  | None -> true
  | Some (Predicate p) -> is_nonexpansive p
  ...
in
...
```

so, maybe let's get rid of `guard_contains_exception_pat` and just do everything in the single pattern match on the guard?

Reply:
makes sense, done!